exchange/bitswap/wantmanager.go

```diff
@@ -0,0 +1,262 @@
+package bitswap
+
+import (
+	"sync"
+	"time"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	engine "github.com/ipfs/go-ipfs/exchange/bitswap/decision"
+	bsmsg "github.com/ipfs/go-ipfs/exchange/bitswap/message"
+	bsnet "github.com/ipfs/go-ipfs/exchange/bitswap/network"
+	wantlist "github.com/ipfs/go-ipfs/exchange/bitswap/wantlist"
+	peer "github.com/ipfs/go-ipfs/p2p/peer"
+	u "github.com/ipfs/go-ipfs/util"
+)
+
+type WantManager struct {
+	receiver bsnet.Receiver
+
+	incoming chan []*bsmsg.Entry
+
+	// notification channel for new peers connecting
+	connect chan peer.ID
+
+	// notification channel for peers disconnecting
+	disconnect chan peer.ID
+
+	peers map[peer.ID]*msgQueue
+
+	wl *wantlist.Wantlist
+
+	network bsnet.BitSwapNetwork
+
+	ctx context.Context
+}
+
+func NewWantManager(ctx context.Context, network bsnet.BitSwapNetwork) *WantManager {
+	return &WantManager{
+		incoming:   make(chan []*bsmsg.Entry, 10),
+		connect:    make(chan peer.ID, 10),
+		disconnect: make(chan peer.ID, 10),
+		peers:      make(map[peer.ID]*msgQueue),
+		wl:         wantlist.New(),
+		network:    network,
+		ctx:        ctx,
+	}
+}
+
+type msgPair struct {
+	to  peer.ID
+	msg bsmsg.BitSwapMessage
+}
+
+type cancellation struct {
+	who peer.ID
+	blk u.Key
+}
+
+type msgQueue struct {
+	p peer.ID
+
+	outlk sync.Mutex
+	out   bsmsg.BitSwapMessage
+
+	work chan struct{}
+	done chan struct{}
+}
+
+func (pm *WantManager) WantBlocks(ks []u.Key) {
+	log.Infof("want blocks: %s", ks)
+	pm.addEntries(ks, false)
+}
+
+func (pm *WantManager) CancelWants(ks []u.Key) {
+	pm.addEntries(ks, true)
+}
+
+func (pm *WantManager) addEntries(ks []u.Key, cancel bool) {
+	var entries []*bsmsg.Entry
+	for i, k := range ks {
+		entries = append(entries, &bsmsg.Entry{
+			Cancel: cancel,
+			Entry: wantlist.Entry{
+				Key:      k,
+				Priority: kMaxPriority - i,
+			},
+		})
+	}
+	select {
+	case pm.incoming <- entries:
+	case <-pm.ctx.Done():
+	}
+}
+
+func (pm *WantManager) SendBlock(ctx context.Context, env *engine.Envelope) {
+	// Blocks need to be sent synchronously to maintain proper backpressure
+	// throughout the network stack
+	defer env.Sent()
+
+	msg := bsmsg.New(false)
+	msg.AddBlock(env.Block)
+	log.Infof("Sending block %s to %s", env.Peer, env.Block)
+	err := pm.network.SendMessage(ctx, env.Peer, msg)
+	if err != nil {
+		log.Error(err)
```

Comment:
This type of thing should NOT be shown to average users. this is what all the levels are for, use NOTICE or INFO if you don't like debug. people can turn it on easily with `IPFS_LOGGING=notice`